File: problems-extend-adapt/0901-1000/0937_log-ordering-rules/solution.cpp
class Solution {
  public:
    vector<string> orderLogs(vector<string> &logs) {
        // Each letter entry carries (content, identifier, original log);
        // digit logs are set aside untouched.
        vector<tuple<string, string, string>> letter;
        vector<string> digit;
        for (const string &log : logs) {
            size_t space = log.find(' ');
            string ident = log.substr(0, space);
            string content = log.substr(space + 1);
            // The content's first character classifies the log: a digit
            // makes it a digit-log, which the sort never touches.
            if (isdigit(static_cast<unsigned char>(content[0]))) {
                digit.push_back(log);
            } else {
                letter.emplace_back(content, ident, log);
            }
        }
        // Letter-logs order by (content, identifier) — tuple comparison is
        // exactly that key pair, and equal keys mean identical logs — then
        // every digit-log follows in its input position.
        stable_sort(letter.begin(), letter.end());
        vector<string> result;
        result.reserve(logs.size());
        for (const auto &[content, ident, log] : letter) {
            result.push_back(log);
        }
        for (const string &log : digit) {
            result.push_back(log);
        }
        return result;
    }
};

File: problems-extend-adapt/0901-1000/0937_log-ordering-rules/solution.py
from typing import List, Optional


class Solution:
    def orderLogs(self, logs: List[str]) -> List[str]:
        # Each letter entry carries (content, identifier, original log);
        # digit logs are set aside untouched.
        letter, digit = [], []
        for log in logs:
            space = log.index(" ")
            ident, content = log[:space], log[space + 1 :]
            # The content's first character classifies the log: a digit
            # makes it a digit-log, which the sort never touches.
            if content[0].isdigit():
                digit.append(log)
            else:
                letter.append((content, ident, log))
        # Letter-logs order by (content, identifier) — a total order, since
        # equal keys mean identical logs — then every digit-log follows in
        # its input position.
        letter.sort(key=lambda entry: (entry[0], entry[1]))
        return [log for _, _, log in letter] + digit
